s3/src/main/scala/akka/stream/alpakka/s3/impl/S3Stream.scala

```diff
@@ -642,53 +602,99 @@ import scala.collection.immutable
       s3Headers: S3Headers,
-      chunkSize: Int
+      chunkSize: Int,
+      maxRetriesPerChunk: Int
   )(parallelism: Int): Flow[ByteString, UploadPartResponse, NotUsed] = {
 
     // Multipart upload requests (except for the completion api) are created here.
     //  The initial upload request gets executed within this function as well.
     //  The individual upload part requests are created.
-    val requestFlow = createRequests(s3Location, contentType, s3Headers, chunkSize, parallelism)
 
-    // The individual upload part requests are processed here
+    assert(
+      chunkSize >= MinChunkSize,
+      s"Chunk size must be at least 5 MB = $MinChunkSize bytes (was $chunkSize bytes). See http://docs.aws.amazon.com/AmazonS3/latest/API/mpUploadUploadPart.html"
+    )
+    assert(maxRetriesPerChunk >= 0, "Max retries per chunk must be >= 0")
+
+    val chunkBufferSize = chunkSize * 2
+
+    // First step of the multi part upload process is made.
+    //  The response is then used to construct the subsequent individual upload part requests
+    val requestInfo: Source[(MultipartUpload, Int), NotUsed] =
+      initiateUpload(s3Location, contentType, s3Headers.headersFor(InitiateMultipartUpload))
+
+    val headers = s3Headers.serverSideEncryption.toIndexedSeq.flatMap(_.headersFor(UploadPart))
+
     Flow
-      .setup { (mat, attr) =>
-        implicit val materializer = mat
-        import mat.executionContext
-        implicit val sys = mat.system
-        implicit val conf = resolveSettings(attr, mat.system)
+      .fromMaterializer { (mat, attr) =>
+        implicit val conf: S3Settings = resolveSettings(attr, mat.system)
+        implicit val sys: ActorSystem = mat.system
+        implicit val materializer: Materializer = mat
+
+        val retriableFlow: Flow[(Chunk, (MultipartUpload, Int)), (Try[HttpResponse], (MultipartUpload, Int)), NotUsed] =
+          Flow[(Chunk, (MultipartUpload, Int))]
+            .map {
+              case (chunkedPayload, (uploadInfo, chunkIndex)) =>
+                //each of the payload requests are created
+                val partRequest =
+                  uploadPartRequest(uploadInfo, chunkIndex, chunkedPayload.data, chunkedPayload.size, headers)
+                (partRequest, (uploadInfo, chunkIndex))
+            }
+            .flatMapConcat { case (req, info) => Signer.signedRequest(req, signingKey).zip(Source.single(info)) }
+            .via(superPool[(MultipartUpload, Int)])
 
-        requestFlow
-          .via(superPool[(MultipartUpload, Int)])
+        SplitAfterSize(chunkSize, chunkBufferSize)(atLeastOneByteString)
+          .via(getChunkBuffer(chunkSize, chunkBufferSize)) //creates the chunks
+          .concatSubstreams
+          .zip(requestInfo)
+          // Allow requests that fail with transient errors to be retried, using the already buffered chunk.
+          .via(RetryFlow.withBackoff(200.millis, 10.seconds, 0.0, maxRetriesPerChunk, retriableFlow) {
```

Comment:
Wasn't sure if it's worth allowing the back-off settings to be configured, and whether they (and the max retries) should be configurable globally or per-upload.  Very open to opinions on this.

Reply:
Done.